### PR TITLE
Allow prefetch_associations to be used on existing records.

### PR DESCRIPTION
**Pull request by dylanahsmith**
@boourns & @eapache for review

## Problem

I would like to be able to prefetch associations on records which have already been loaded.

An example of where this can be useful is when performing an update on a record.  The record needs to be loaded from the database to avoid propagating cache corruption, but that prevents prefetching unloaded associations using the existing record when creating an HTTP response that corresponds to that update.

This can also be useful for normalized associations, like when calling `fetch_#{association}` on a `cache_belongs_to` association or on an id embedded association, which doesn't support the `:includes` option, since it would complicate memorization.

## Solution

I made the `prefetch_associations` method public and made it work with records not loaded from IDC.  Embedded associations are loaded by fetching the owner of the association from the cache, then moving the loaded association records onto the records passed to `prefetch_associations`.

### CODE REVIEW

File: lib/identity_cache/query_api.rb [hunk 1]
```diff
@@ -246,96 +272,102 @@ def find_batch(ids)
...
 
-              parent_record_to_child_records = Hash.new { |h, k| h[k] = [] }
-              child_records = details[:association_reflection].klass.fetch_multi(*ids_to_parent_record.keys)
-              child_records.each do |child_record|
-                parent_record = ids_to_parent_record[child_record.id]
-                parent_record_to_child_records[parent_record] << child_record
-              end
+      def prefetch_embedded_association(records, association, details)
+        first_record = records.first
+        return if first_record.association(association).loaded?
```
Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: This is the same assumption that active record makes in `ActiveRecord::Associations::Preloader`

File: lib/identity_cache/configuration_dsl.rb [hunk 2]
```diff
@@ -201,17 +201,14 @@ def build_recursive_association_cache(association, options) #:nodoc:
         options[:cached_accessor_name]   = "fetch_#{association}"
         options[:records_variable_name]  = "cached_#{association}"
 
+        self.class_eval(<<-CODE, __FILE__, __LINE__ + 1)
+          def #{options[:cached_accessor_name]}
+            fetch_recursively_cached_association('#{options[:records_variable_name]}', :#{association})
+          end
+        CODE
 
-        unless instance_methods.include?(options[:cached_accessor_name].to_sym)
```
Comment (eapache) [inline lib/identity_cache/configuration_dsl.rb]: this change doesn't seem strictly related?

Comment (dylanahsmith) [inline lib/identity_cache/configuration_dsl.rb]: Right, this was a useless check, but I could make a separate PR for it.

**Comment by boourns** [comment]
LGTM

File: lib/identity_cache/configuration_dsl.rb (hunk 2 above)
Comment (eapache) [inline lib/identity_cache/configuration_dsl.rb]: No need, I just thought I was missing something.

File: lib/identity_cache/query_api.rb [hunk 3]
```diff
@@ -246,96 +272,102 @@ def find_batch(ids)
...
+        iv_name_key = details[:embed] == true ? :records_variable_name : :ids_variable_name
+        return if first_record.instance_variable_get(:"@#{details[iv_name_key]}")
+        fetch_embedded_associations(records)
+      end
 
-              parent_record_to_child_records.each do |parent_record, child_records|
-                parent_record.send(details[:prepopulate_method_name], child_records)
+      def prefetch_one_association(association, records)
```
Comment (eapache) [inline lib/identity_cache/query_api.rb]: this method is enormous... can we split the three cases into their own methods for readability?

File: lib/identity_cache/query_api.rb [hunk 4]
```diff
@@ -74,6 +74,32 @@ def fetch_multi(*ids)
         records
       end
 
+      def prefetch_associations(associations, records)
+        return if records.empty?
+
+        case associations
+        when nil
```
Comment (eapache) [inline lib/identity_cache/query_api.rb]: this is an intentional do-nothing case? would be nice to have a comment making that explicit, otherwise people might think it was supposed to fall-through

Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: Yup, we supported that previously with the `when nil` in hashify_includes_structure.  I'll add a comment.

File: lib/identity_cache/query_api.rb (hunk 1 above)
Comment (eapache) [inline lib/identity_cache/query_api.rb]: put this as an in-code comment please

File: lib/identity_cache/query_api.rb [hunk 5]
```diff
@@ -246,96 +272,102 @@ def find_batch(ids)
         ids.map{ |id| records_by_id[id] }
       end
 
-      def prefetch_associations(associations, records)
-        associations = hashify_includes_structure(associations)
+      def fetch_embedded_associations(records)
+        associations = embedded_associations
+        return true if associations.empty?
```
Comment (eapache) [inline lib/identity_cache/query_api.rb]: AFAICT nothing is using these return values?

File: lib/identity_cache/query_api.rb [hunk 6]
```diff
@@ -246,96 +272,102 @@ def find_batch(ids)
...
 
-        associations.each do |association, sub_associations|
-          case
-          when details = cached_has_manys[association]
+        return false unless primary_cache_index_enabled
 
-            if details[:embed] == true
-              child_records = records.map(&details[:cached_accessor_name].to_sym).flatten
+        records_by_id = records.index_by(&:id)
+        cached_records = fetch_multi(records.map(&:id))
```
Comment (eapache) [inline lib/identity_cache/query_api.rb]: not sure this line makes sense in the context of the change - isn't the whole point that if we were passed in a complete record already (e.g. one fetched straight from the DB in an update scenario) that we *don't* want to go to the cache for that record because it could have incorrect associations, and we just want to go the cache for the associations of the "concrete" record we already have?

File: lib/identity_cache/query_api.rb (hunk 3 above)
Comment (eapache) [inline lib/identity_cache/query_api.rb]: Oh, I see when reviewing with `?w=0` that this method actually didn't really change, just got unindented

File: lib/identity_cache/query_api.rb (hunk 6 above)
Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: The whole point of this method *is* to go the cache for the associations.  However, those associations are embedded on a record, so we need to fetch the record to get at the associations.

I don't know what you mean about *incorrect associations*.  In relation to a mutation, the cache would have been invalidated to avoid fetching anything incorrect.

Comment (eapache) [inline lib/identity_cache/query_api.rb]: > The whole point of this method is to go the cache for the associations.

Yes.

> However, those associations are embedded on a record, so we need to fetch the record to get at the associations.

We already have the record; that's the point of this PR, to load associations when we already have th record.

Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: I don't really understand what is the problem with this line then, it seems essential to this pull request.

File: lib/identity_cache/query_api.rb [hunk 7]
```diff
@@ -246,96 +272,102 @@ def find_batch(ids)
...
+        return true if associations.empty?
 
-        associations.each do |association, sub_associations|
-          case
-          when details = cached_has_manys[association]
+        return false unless primary_cache_index_enabled
 
-            if details[:embed] == true
-              child_records = records.map(&details[:cached_accessor_name].to_sym).flatten
+        records_by_id = records.index_by(&:id)
```
Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: Actually, this assumes `records` are unique, which could be avoided by indexing the cached records.

File: lib/identity_cache/query_api.rb (hunk 5 above)
Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: Oh right.  I wasn't quite sure how to handle the case where the parent record has been deleted and can't be refetched.  In many cases the associations have also been deleted, but it is probably safer to fallback to the database for this edge case.  We probably don't need to worry about efficiency for this edge case, so I never ended up using the result.

I'll change these to not explicitly return anything.

File: lib/identity_cache/query_api.rb (hunk 4 above)
Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: `# do nothing` was added

File: lib/identity_cache/query_api.rb (hunk 6 above)
Comment (eapache) [inline lib/identity_cache/query_api.rb]: We already have the record(s). Why are we re-fetching the record(s)?

Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: We have the records, but not their embedded associations.  How can you fetch the embedded associations from the cache?  The only place they are stored in the cache is in the same cache entry that would store these records that we already have.

**Comment by eapache** [comment]
LGTM
